### PR TITLE
Use parameter broadcasting in `KerasLayer` and `TorchLayer`

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
The `KerasLayer` and `TorchLayer` classes have yet to be updated to make use of parameter broadcasting. The current behaviour is to unstack the input along the batch dimension, then execute the QNode once for each element in the batch. This leads to very slow execution times as detailed in [this](https://discuss.pennylane.ai/t/backward-function-takes-long-and-batches/164/4) forum post. 

**Description of the Change:**
Remove the unstacking logic and pass the full (batched) input directly to the QNode. This makes of parameter broadcasting inside PennyLane.

Also add an additional keyword argument `split_batches` to these two classes which enables the old behaviour. This is necessary because some components (e.g. density matrices) do not support parameter broadcasting yet, and some of these components may be used in ML models.

**Benefits:**
Faster training for models that use `KerasLayer` and `TorchLayer`.

**Possible Drawbacks:**
The keyword argument might be confusing (however its usage is documented).

**Related GitHub Issues:**
https://discuss.pennylane.ai/t/backward-function-takes-long-and-batches/164/4

### CODE REVIEW

**Comment by eddddddy** [comment]
[sc-36425]

File: pennylane/qnn/keras.py [hunk 1]
```diff
@@ -54,10 +51,9 @@ class KerasLayer(Layer):
             arguments of the `add_weight()
             <https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#add_weight>`__
             method and values being the corresponding specification.
-        batch_idx (Union[Sequence[int], int]): Argument location of the non-trainable inputs for
-            the circuit. This allows batch execution by creating executable circuits for each
-            input example with the same trainable weights. Default ``None``.
-            See :func:`~.pennylane.transforms.batch_input` for more details.
+        split_batches (bool): If True, any batch dimension in the input tensor will be unstacked
+            and split into multiple QNode executions. Otherwise (default), internal PennyLane
+            broadcasting will be used.
```
Comment (trbromley) [inline pennylane/qnn/keras.py:56]: What's the use case for `split_batches=True`?

Comment (eddddddy) [reply pennylane/qnn/keras.py:56]: Two main reasons to allow the old behaviour:
1. There are some components of PennyLane (e.g. `qml.math.density_matrix`) that don't support parameter broadcasting yet. Users may still want these components in their ML models.
2. Parameter broadcasting doesn't support more than one batch dimension. The old behaviour supports more than one batch dimension.

**Review by dwierichs — COMMENTED** [review]
Nice, thanks for this @eddddddy 
I'll give it a second pass tomorrow to make sure I really understand the reshaping steps (they feel a bit orthogonal to what the new return type system does :D )
Could you also add a changelog entry, in particular because there are breaking changes?

File: pennylane/qnn/torch.py [hunk 2]
```diff
@@ -302,14 +308,14 @@ def forward(self, inputs):  # pylint: disable=arguments-differ
...
-        if len(inputs.shape) > 1:
+        if self.split_batches and len(inputs.shape) > 1:
             # If the input size is not 1-dimensional, unstack the input along its first dimension,
             # recursively call the forward pass on each of the yielded tensors, and then stack the
             # outputs back into the correct shape
             reconstructor = [self.forward(x) for x in torch.unbind(inputs)]
             return torch.stack(reconstructor)
 
-        # If the input is 1-dimensional, calculate the forward pass as usual
+        # calculate the forward pass as usual
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/qnn/torch.py:318]: The next lines are executed if either i) the input is 1-dimensional, or ii) `split_batches` is `False`, in which case it will make use of parameter broadcasting. I can add a comment to make this explicit.

[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/qnn/torch.py:318]: Ah, I see. Not a blocker from my side, but when in doubt, giving helpful comments doesn't hurt :)

File: pennylane/qnn/keras.py [hunk 3]
```diff
@@ -293,14 +286,14 @@ def call(self, inputs):
         Returns:
             tensor: output data
         """
-        if len(tf.shape(inputs)) > 1 and self.argnum is None:
+        if self.split_batches and len(tf.shape(inputs)) > 1:
```
Comment (dwierichs) [inline pennylane/qnn/keras.py:289]: I'm wondering: Given that we do not support more than one parameter broadcasting axis, one could also iteratively unstack in case there are more than one axis, no?
Along the lines of 
```suggestion
        if ((ndim:=len(tf.shape(inputs))) > 1 and self.split_batches) or ndim>2:
```

File: pennylane/qnn/keras.py [hunk 4]
```diff
@@ -318,8 +311,14 @@ def _evaluate_qnode(self, x):
             **{k: 1.0 * w for k, w in self.qnode_weights.items()},
         }
         res = self.qnode(**kwargs)
+
         if isinstance(res, (list, tuple)):
+            if len(tf.shape(x)) > 1:
+                # multi-return and batch dim case
+                res = [tf.reshape(r, (tf.shape(x)[0], -1)) for r in res]
```
Comment (dwierichs) [inline pennylane/qnn/keras.py:318]: So for mixed return types, this step concatenates them into a flat array, do I understand this correctly?
This approach differs from the (new) return type system of PennyLane, do we care about this fact? I am not too familiar with keras layers, but I suppose it is required that it returns a flat output dimension.

File: pennylane/qnn/torch.py [hunk 5]
```diff
@@ -302,14 +308,14 @@ def forward(self, inputs):  # pylint: disable=arguments-differ
             tensor: output data
         """
 
-        if len(inputs.shape) > 1:
+        if self.split_batches and len(inputs.shape) > 1:
```
Comment (dwierichs) [inline pennylane/qnn/torch.py:311]: Same as for keras regarding multiple batch dimensions.

File: pennylane/qnn/keras.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/qnn/keras.py:318]: To be clear: This is not a change made by this PR, the `hstack` was there before, so I'm just curious whether we are aware of this inconsistency. :)

File: pennylane/qnn/torch.py [hunk 6]
```diff
@@ -395,3 +404,13 @@ def __str__(self):
...
+
+    @staticmethod
+    def set_input_argument(input_name: Text = "inputs") -> None:
+        """
+        Set the name of the input argument.
+
+        Args:
+            input_name (str): Name of the input argument
+        """
+        TorchLayer._input_arg = input_name
```
Comment (dwierichs) [inline pennylane/qnn/torch.py:416]: Why is this needed? Is it related to broadcasting?

**Review by dwierichs — COMMENTED** [review]
Found a few more small questions in the second iteration (see comments for details):
1. Do we want to auto-dispatch to unstacking inputs whenever there are multiple broadcasting axes?
2. Is the (already existing) way of stacking QNode outputs into a flat array still the way we want to go? It seems inconsistent with the new return type system.
3. Why do we need `set_input_argument`?

File: pennylane/qnn/keras.py (hunk 1 above)
Comment (rmoyard) [reply pennylane/qnn/keras.py:56]: Once the density_matrix is updated, I would suggest to remove this.

File: pennylane/qnn/keras.py (hunk 3 above)
Comment (rmoyard) [reply pennylane/qnn/keras.py:289]: If we want to batch for more than one dimension, let's make the different cases very clear in the documentation because it will become hard for users to chose the right configuration.

File: pennylane/qnn/keras.py (hunk 4 above)
Comment (rmoyard) [reply pennylane/qnn/keras.py:318]: Yes I am not sure you can return anything else than a single dimension tf tensor here?

File: pennylane/qnn/torch.py (hunk 6 above)
Comment (rmoyard) [reply pennylane/qnn/torch.py:416]: Same question, where is it called?

**Review by rmoyard — COMMENTED** [review]
It looks good @eddddddy, just a couple of questions before approving :+1:

File: pennylane/qnn/torch.py (hunk 6 above)
Comment (eddddddy) [reply pennylane/qnn/torch.py:416]: It's not related to broadcasting. The same static method is present in `KerasLayer`, which allows the user to set the name of the input argument to the QNode that needs to be broadcasted. It defaults to `"inputs"`.

File: pennylane/qnn/keras.py (hunk 4 above)
Comment (eddddddy) [reply pennylane/qnn/keras.py:318]: The resulting tensor (after `hstack`ing) has shape `(batch_size, output_dim)`. This is consistent with what we expect with parameter broadcasting.

I believe the previous implementation (with `hstack` just by itself) has a bug. If, for example, the user passes in a batch size of 4 and the QNode has two `expval` measurements, then the previous implementation would return an array of shape `(8,)` instead of `(4, 2)`.

File: pennylane/qnn/keras.py (hunk 3 above)
Comment (eddddddy) [reply pennylane/qnn/keras.py:289]: Oops, I think there's a much simpler solution by reshaping the arrays so that there's only one batch dimension, then reshaping them back at the end.

File: pennylane/qnn/keras.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/qnn/keras.py:56]: I'm not sure whether there are concrete plans to do this currently :thinking:

File: pennylane/qnn/keras.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/qnn/keras.py:318]: In the return types system, this would lead to a `tuple` being returned, but I guess that is not really desired when working with a `Layer` as part of a QNN...

File: pennylane/qnn/torch.py (hunk 6 above)
Comment (dwierichs) [reply pennylane/qnn/torch.py:416]: Cool :)

File: pennylane/qnn/keras.py (hunk 4 above)
Comment (eddddddy) [reply pennylane/qnn/keras.py:331]: Nope, we definitely want to return a single `tf.Tensor`, but it's still up in the air what the behaviour should be when combining measurements with incompatible shapes.

File: pennylane/qnn/keras.py (hunk 3 above)
Comment (eddddddy) [reply pennylane/qnn/keras.py]: At least the case of more than one broadcasting axis is solved. We will need the `split_batches` argument until we decide what to do with density matrices.

File: pennylane/qnn/torch.py (hunk 5 above)
Comment (eddddddy) [reply pennylane/qnn/torch.py]: https://github.com/PennyLaneAI/pennylane/pull/4131#discussion_r1196678630

File: pennylane/qnn/keras.py (hunk 1 above)
Comment (eddddddy) [reply pennylane/qnn/keras.py:56]: In that case, alternatively we can check if the QNode has a `density_matrix` measurement, and use the old behaviour if it does. The implementation will be a bit messier but at least it'll be less confusing for the user.

File: pennylane/qnn/torch.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/qnn/torch.py]: Added some more comments 👍

File: pennylane/qnn/keras.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/qnn/keras.py:59]: It should definitely be the responsibility of the device to choose whether or not it has to split a tape with broadcasted parameters up into multiple tapes. 

I wasn't actually aware of this, and I would call it a bug that needs to be fixed. We either need to make density matrix measurements support parameter broadcasting or split up the tape in `Device.batch_transform` if a measurement is a `density_matrix`.

Comment (albi3ro) [reply pennylane/qnn/keras.py:59]: Once the new transform infrastructure comes in, users can just apply the relevant transform directly to the QNode, instead of needing a keyword argument here.

File: pennylane/qnn/keras.py (hunk 4 above)
Comment (rmoyard) [reply pennylane/qnn/keras.py:362]: I agree with you @eddddddy, that would be a good solution, we cannot have the return system from PennyLane there. This post processing makes sense because we are using external libraries and have to adapt to their need.

Comment (dwierichs) [reply pennylane/qnn/keras.py:362]: This sounds very reasonable to me. :+1:

File: pennylane/qnn/keras.py (hunk 1 above)
Comment (eddddddy) [reply pennylane/qnn/keras.py:59]: Removed this keyword argument and all broadcasting logic out of these classes.

File: pennylane/qnn/keras.py (hunk 4 above)
Comment (eddddddy) [reply pennylane/qnn/keras.py:375]: Changed and added docs.

**Review by rmoyard — APPROVED** [review]
Thanks a lot for adding the docs for the different shapes! It looks good to me :100:
